### ISSUE TITLE
Proper deletion of FlatStateDeltas

### ISSUE THREAD

**Issue by Longarithm** [near/nearcore#8655]
We've discovered that FlatStateDeltas size is 200MB on live mainnet node, though only 20 of them are actively being changed - which means that we forget to remove some deltas, probably during migration.

Also DB ops trace show that there are more deletions than insertions https://github.com/near/nearcore/issues/8006#issuecomment-1448183752

**Comment by jakmeier** [comment]
Here is an excerpt of FlatStateDeltas operations that shows double deleting:

```bash
  SET FlatStateDeltas "AAAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw==" size=907  # key AA..gw is inserted
  SET FlatStateDeltas "AQAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw==" size=4
  SET FlatStateDeltas "AgAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw==" size=250
  SET FlatStateDeltas "AwAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw==" size=759
...
  DELETE FlatStateDeltas "AAAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw==" # AA..gw  is delete here
  DELETE FlatStateDeltas "AAAAAAAAAACtYa7RPuC3gP6qYBvIvfrpfWG9mAOgjL1Ei7sq0Czcog=="
  DELETE FlatStateDeltas "AQAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw=="
  DELETE FlatStateDeltas "AQAAAAAAAACtYa7RPuC3gP6qYBvIvfrpfWG9mAOgjL1Ei7sq0Czcog=="
  DELETE FlatStateDeltas "AgAAAAAAAACtYa7RPuC3gP6qYBvIvfrpfWG9mAOgjL1Ei7sq0Czcog=="
  DELETE FlatStateDeltas "AgAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw=="
  DELETE FlatStateDeltas "AwAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw=="
  DELETE FlatStateDeltas "AwAAAAAAAACtYa7RPuC3gP6qYBvIvfrpfWG9mAOgjL1Ei7sq0Czcog=="
  SET FlatStateDeltas "AAAAAAAAAADndU2QiKg0jLm7KbvEEs0MBkL8eZ7n8E+PCj4vRWvtBw==" size=1708
  SET FlatStateDeltas "AQAAAAAAAADndU2QiKg0jLm7KbvEEs0MBkL8eZ7n8E+PCj4vRWvtBw==" size=4
  SET FlatStateDeltas "AgAAAAAAAADndU2QiKg0jLm7KbvEEs0MBkL8eZ7n8E+PCj4vRWvtBw==" size=110
  SET FlatStateDeltas "AwAAAAAAAADndU2QiKg0jLm7KbvEEs0MBkL8eZ7n8E+PCj4vRWvtBw==" size=624
  DELETE FlatStateDeltas "AAAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw=="  # AA..gw  is delete again here
  DELETE FlatStateDeltas "AAAAAAAAAAAr9sDRXk4UkL1R6Kf0vJZ8KKp8tk8Bs6IUyirLXAm3Aw=="
  DELETE FlatStateDeltas "AQAAAAAAAAAr9sDRXk4UkL1R6Kf0vJZ8KKp8tk8Bs6IUyirLXAm3Aw=="
  DELETE FlatStateDeltas "AQAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw=="
  DELETE FlatStateDeltas "AgAAAAAAAAAr9sDRXk4UkL1R6Kf0vJZ8KKp8tk8Bs6IUyirLXAm3Aw=="
  DELETE FlatStateDeltas "AgAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw=="
  DELETE FlatStateDeltas "AwAAAAAAAAChhHCJ7mSN2T/ydoZBuNstNpiXedFgBWq2hGB+3BhIgw=="
```

I think all keys are deleted exactly twice.

**Comment by jakmeier** [comment]
I just confirmed the duplicate deletion running on git revision `229e6ae763089c7b9a14062652e8fbbcfd35eb62`, the exact hash of master where the FS cache was merged.

Again, I see all keys deleted twice. But it's hard to give full context in a snippet. The second deletion actually happens many DB commits later, so in the full IO trace they are a few thousand lines apart. (The previous excerpt only showed all Delta ops.)

If you want to look into the exact trace, here is a relatively short one. 
[20230228_229e6ae_mainnet.iotrace.zip](https://github.com/near/nearcore/files/10853202/20230228_229e6ae_mainnet.iotrace.zip)
